src/passes/OptimizeInstructions.cpp

```diff
@@ -1391,6 +1391,20 @@ struct OptimizeInstructions
       right->type = Type::i32;
       return right;
     }
+    {
+      Binary* inner;
+      Expression* right;
+      // ~(1 << x)   ==>   (1 << x) ^ -1  ==>  rotl(-2, x)
+      if (matches(curr,
+                  binary(Abstract::Xor,
+                         binary(&inner, Abstract::Shl, ival(1), any(&right)),
+                         ival(-1)))) {
+        inner->op = Abstract::getBinary(type, Abstract::RotL);
+        inner->left = LiteralUtils::makeFromInt32(-2, type, *getModule());
+        inner->right = right;
+        return inner;
```

Comment:
You could avoid having to extract `inner` by reusing `curr` instead. It would also be good to rename this local `right` variable to `x` to make the pattern more compact and to match the comment more exactly.

Otherwise LGTM!